File: src/arbitrated_pool/state.ts
import Transaction from '@bsv/sdk/transaction/Transaction';
import Script from '@bsv/sdk/script/Script';
import LockingScript from '@bsv/sdk/script/LockingScript';
import UnlockingScript from '@bsv/sdk/script/UnlockingScript';
import { buildOptionalOpReturnScript } from '../libs/OP_RETURN';
import { estimateSerializedTxSize } from '../libs/TX_SIZE';
import { Protocol, validatePoolProtocol, Version, type ArbitratedPoolRoles, type ArbitratedPoolStateInput } from '../types';
import {
  buildArbitratedPoolLock,
  buildArbitratedPoolOutputScripts,
  cloneArbitratedPoolTransaction,
  requireSource,
  validateArbitratedPoolRoles,
  validateArbitratedPoolStateOutputs,
} from './script';

const MAX_SAFE = Number.MAX_SAFE_INTEGER;

function empty(): UnlockingScript { return new UnlockingScript(); }

function fake(): UnlockingScript {
  const script = new Script([]);
  script.writeOpCode(0);
  script.writeBin(new Array(73).fill(0));
  script.writeBin(new Array(73).fill(0));
  const unlocking = new UnlockingScript();
  unlocking.chunks = script.chunks;
  return unlocking;
}

function assertSafeInteger(name: string, value: unknown): asserts value is number {
  if (!Number.isSafeInteger(value)) throw new Error(`${name} must be a safe integer`);
}

function assertAmount(name: string, value: unknown, positive = false): asserts value is number {
  assertSafeInteger(name, value);
  if ((positive && value <= 0) || (!positive && value < 0)) throw new Error(`${name} must not be negative`);
}

function assertStateInput(input: ArbitratedPoolStateInput): void {
  validatePoolProtocol(input);
  assertAmount('pool amount', input.poolAmount, true);
  assertAmount('seller amount', input.sellerAmount);
  assertAmount('arbiter amount', input.arbiterAmount);
  assertSafeInteger('sequence', input.sequence);
  if (input.sequence < 0 || input.sequence > 0xffffffff) throw new Error('sequence must be a uint32');
  assertSafeInteger('fee rate', input.feeRate);
  if (input.feeRate < 0) throw new Error('fee rate must not be negative');
  if (input.lockTime !== undefined) {
    assertSafeInteger('lock time', input.lockTime);
    if (input.lockTime < 0 || input.lockTime > 0xffffffff) throw new Error('lock time must be a uint32');
  }
  if (input.buyerAmount !== undefined) assertAmount('buyer amount', input.buyerAmount);
  validateArbitratedPoolRoles(input.roles);
}

function checkedAllocation(sellerAmount: number, arbiterAmount: number, poolAmount: number): number {
  if (sellerAmount > MAX_SAFE - arbiterAmount) throw new Error('allocated amount overflow');
  const allocatedAmount = sellerAmount + arbiterAmount;
  if (allocatedAmount > poolAmount) throw new Error('allocated amount exceeds pool amount');
  return allocatedAmount;
}

function checkedFee(size: number, rate: number): number {
  if (size < 0 || !Number.isSafeInteger(size)) throw new Error('serialized transaction size is invalid');
  if (rate === 0) return 0;
  if (size > Math.floor((MAX_SAFE - 999) / rate)) throw new Error('transaction fee overflow');
  const product = size * rate;
  if (product > MAX_SAFE - 999) throw new Error('transaction fee overflow');
  return Math.floor((product + 999) / 1000);
}

function clone(tx: Transaction): Transaction { return cloneArbitratedPoolTransaction(tx); }

export async function buildArbitratedPoolState(input: ArbitratedPoolStateInput): Promise<Transaction> {
  assertStateInput(input);
  if (!input.previousState) throw new Error('previous state is required');
  const state = clone(input.previousState);
  if (state.inputs.length !== 1 || !state.inputs[0]) throw new Error('arbitrated pool state must have exactly one input');
  validateArbitratedPoolStateOutputs(state, input.roles);
  if (input.sequence <= (state.inputs[0].sequence ?? 0)) throw new Error('payment sequence must increase');

  const lock = buildArbitratedPoolLock(input.roles);
  requireSource(state, input.poolAmount, lock);
  const outputScripts = buildArbitratedPoolOutputScripts(input.roles);
  const allocatedAmount = checkedAllocation(input.sellerAmount, input.arbiterAmount, input.poolAmount);
  const buyerBeforeFee = input.poolAmount - allocatedAmount;

  state.outputs[0].lockingScript = outputScripts.buyer;
  state.outputs[1].lockingScript = outputScripts.seller;
  state.outputs[2].lockingScript = outputScripts.arbiter;
  state.outputs[0].satoshis = buyerBeforeFee;
  state.outputs[1].satoshis = input.sellerAmount;
  state.outputs[2].satoshis = input.arbiterAmount;
  state.inputs[0].sequence = input.sequence;
  if (input.lockTime !== undefined) state.lockTime = input.lockTime;

  const proof = buildOptionalOpReturnScript(input.paymentProof);
  if (proof) {
    const outputScript = new LockingScript();
    outputScript.chunks = proof.chunks;
    if (state.outputs.length === 4) state.outputs[3] = { lockingScript: outputScript, satoshis: 0 };
    else state.addOutput({ lockingScript: outputScript, satoshis: 0 });
  }

  state.inputs[0].unlockingScript = fake();
  const fee = checkedFee(estimateSerializedTxSize(state), input.feeRate);
  if (fee > state.outputs[0].satoshis) throw new Error('buyer balance is insufficient for fee');
  state.outputs[0].satoshis -= fee;
  if (input.buyerAmount !== undefined && input.buyerAmount !== state.outputs[0].satoshis) throw new Error('buyer amount does not match canonical fee');
  state.inputs[0].unlockingScript = empty();
  return state;
}

export function buildArbitratedPoolOpeningState(fundingTx: Transaction, poolAmount: number, roles: ArbitratedPoolRoles, lockTime: number, feeRate: number): Promise<Transaction> {
  if (!fundingTx || fundingTx.outputs.length === 0) throw new Error('funding transaction is required');
  const state = new Transaction();
  state.addInput({ sourceTransaction: fundingTx, sourceTXID: fundingTx.id('hex'), sourceOutputIndex: 0, sequence: 1, unlockingScript: empty() });
  const outputScripts = buildArbitratedPoolOutputScripts(roles);
  state.addOutput({ lockingScript: outputScripts.buyer, satoshis: poolAmount });
  state.addOutput({ lockingScript: outputScripts.seller, satoshis: 0 });
  state.addOutput({ lockingScript: outputScripts.arbiter, satoshis: 0 });
  state.lockTime = lockTime;
  return buildArbitratedPoolState({ protocol: Protocol, version: Version, previousState: state, sequence: 2, sellerAmount: 0, arbiterAmount: 0, poolAmount, roles, feeRate });
}

export function buildArbitratedPoolFinalState(input: ArbitratedPoolStateInput): Promise<Transaction> { return buildArbitratedPoolState(input); }

File: pkg/arbitrated_pool/state.go
package arbitrated_pool

import (
	"bytes"
	"fmt"
	"github.com/bsv-blockchain/go-sdk/chainhash"
	"github.com/bsv-blockchain/go-sdk/script"
	tx "github.com/bsv-blockchain/go-sdk/transaction"
	"github.com/bsv-blockchain/go-sdk/transaction/template/p2pkh"
	libs "github.com/bsv8/MultisigPool/v3/pkg/libs"
)

type StateInput struct {
	Protocol      string
	Version       uint32
	PreviousRawTx []byte
	// PreviousSourceOutput 在标准 raw transaction 不包含源输出元数据时显式提供源输出。
	PreviousSourceOutput *tx.TransactionOutput
	Sequence             uint32
	LockTime             *uint32
	BuyerAmount          uint64
	SellerAmount         uint64
	PoolAmount           uint64
	Roles                ArbitratedPoolRoles
	FeeRate              FeeSatPerKB
	PaymentProof         []byte
}

func clone(value *tx.Transaction) (*tx.Transaction, error) {
	if value == nil {
		return nil, fmt.Errorf("transaction is required")
	}
	copy, err := tx.NewTransactionFromBytes(value.Bytes())
	if err != nil {
		return nil, fmt.Errorf("copy transaction: %w", err)
	}
	return copy, nil
}

func scripts(roles ArbitratedPoolRoles) (*script.Script, *script.Script, *script.Script, error) {
	if err := validateRoles(roles); err != nil {
		return nil, nil, nil, err
	}
	lock, err := BuildArbitratedPoolLock(roles)
	if err != nil {
		return nil, nil, nil, err
	}
	buyerAddress, err := libs.GetAddressFromPublicKey(roles.Buyer, false)
	if err != nil {
		return nil, nil, nil, err
	}
	sellerAddress, err := libs.GetAddressFromPublicKey(roles.Seller, false)
	if err != nil {
		return nil, nil, nil, err
	}
	buyer, err := p2pkh.Lock(buyerAddress)
	if err != nil {
		return nil, nil, nil, err
	}
	seller, err := p2pkh.Lock(sellerAddress)
	if err != nil {
		return nil, nil, nil, err
	}
	return lock, buyer, seller, nil
}

// BuildArbitratedPoolState 固定输出 [Buyer, Seller]，Arbiter 没有资金输出。
func BuildArbitratedPoolState(input StateInput) (*tx.Transaction, error) {
	if input.Protocol != Protocol || input.Version != Version {
		return nil, fmt.Errorf("unsupported pool protocol: expected %s v%d", Protocol, Version)
	}
	if len(input.PreviousRawTx) == 0 || input.PoolAmount == 0 {
		return nil, fmt.Errorf("previous state and pool amount are required")
	}
	state, err := tx.NewTransactionFromBytes(input.PreviousRawTx)
	if err != nil {
		return nil, fmt.Errorf("decode previous state: %w", err)
	}
	if len(state.Inputs) != 1 || (len(state.Outputs) != 2 && len(state.Outputs) != 3) || state.Inputs[0] == nil {
		return nil, fmt.Errorf("arbitrated pool state must have one input and two value outputs")
	}
	if input.Sequence <= state.Inputs[0].SequenceNumber {
		return nil, fmt.Errorf("payment sequence must increase")
	}
	lock, buyer, seller, err := scripts(input.Roles)
	if err != nil {
		return nil, err
	}
	source := state.Inputs[0].SourceTxOutput()
	if input.PreviousSourceOutput != nil {
		if source != nil && (source.Satoshis != input.PreviousSourceOutput.Satoshis || !bytes.Equal(source.LockingScript.Bytes(), input.PreviousSourceOutput.LockingScript.Bytes())) {
			return nil, fmt.Errorf("previous state source output does not match configured pool")
		}
		if source == nil {
			source = input.PreviousSourceOutput
		}
	}
	if source == nil {
		return nil, fmt.Errorf("previous state source output is required")
	}
	if source == nil || source.Satoshis != input.PoolAmount || !bytes.Equal(source.LockingScript.Bytes(), lock.Bytes()) {
		return nil, fmt.Errorf("previous state source output does not match configured pool")
	}
	if state.Inputs[0].SourceTxOutput() == nil {
		state.Inputs[0].SetSourceTxOutput(source)
	}
	if !bytes.Equal(state.Outputs[0].LockingScript.Bytes(), buyer.Bytes()) || !bytes.Equal(state.Outputs[1].LockingScript.Bytes(), seller.Bytes()) {
		return nil, fmt.Errorf("previous state outputs do not match buyer and seller roles")
	}
	if input.SellerAmount > input.PoolAmount {
		return nil, fmt.Errorf("seller amount exceeds pool amount")
	}
	state.Outputs[0].Satoshis = input.PoolAmount - input.SellerAmount
	state.Outputs[1].Satoshis = input.SellerAmount
	state.Outputs[0].LockingScript = buyer
	state.Outputs[1].LockingScript = seller
	state.Inputs[0].SequenceNumber = input.Sequence
	if input.LockTime != nil {
		state.LockTime = *input.LockTime
	}
	if len(input.PaymentProof) > 0 {
		proof, err := libs.BuildOptionalOpReturnScript(input.PaymentProof)
		if err != nil {
			return nil, err
		}
		if len(state.Outputs) == 3 {
			state.Outputs[2] = &tx.TransactionOutput{Satoshis: 0, LockingScript: proof}
		} else {
			state.AddOutput(&tx.TransactionOutput{Satoshis: 0, LockingScript: proof})
		}
	}
	fake, err := libs.FakeSign(2)
	if err != nil {
		return nil, err
	}
	state.Inputs[0].UnlockingScript = fake
	fee, err := feeSat(state.Size(), input.FeeRate)
	if err != nil {
		return nil, err
	}
	if fee > state.Outputs[0].Satoshis {
		return nil, fmt.Errorf("buyer balance is insufficient for fee")
	}
	state.Outputs[0].Satoshis -= fee
	if input.BuyerAmount != 0 && input.BuyerAmount != state.Outputs[0].Satoshis {
		return nil, fmt.Errorf("buyer amount does not match canonical fee")
	}
	state.Inputs[0].UnlockingScript = script.NewFromBytes(nil)
	return state, nil
}

func BuildArbitratedPoolOpeningState(fundingTxID []byte, poolOutputIndex uint32, poolAmount uint64, roles ArbitratedPoolRoles, lockTime uint32, rate FeeSatPerKB) (*tx.Transaction, error) {
	if len(fundingTxID) != 32 || poolAmount == 0 {
		return nil, fmt.Errorf("funding outpoint and pool amount are required")
	}
	lock, buyer, seller, err := scripts(roles)
	if err != nil {
		return nil, err
	}
	state := tx.NewTransaction()
	id, err := chainhash.NewHash(fundingTxID)
	if err != nil {
		return nil, err
	}
	state.AddInputWithOutput(&tx.TransactionInput{SourceTXID: id, SourceTxOutIndex: poolOutputIndex, SequenceNumber: 2, UnlockingScript: script.NewFromBytes(nil)}, &tx.TransactionOutput{Satoshis: poolAmount, LockingScript: lock})
	state.AddOutput(&tx.TransactionOutput{Satoshis: poolAmount, LockingScript: buyer})
	state.AddOutput(&tx.TransactionOutput{Satoshis: 0, LockingScript: seller})
	state.LockTime = lockTime
	fake, err := libs.FakeSign(2)
	if err != nil {
		return nil, err
	}
	state.Inputs[0].UnlockingScript = fake
	fee, err := feeSat(state.Size(), rate)
	if err != nil {
		return nil, err
	}
	if fee > poolAmount {
		return nil, fmt.Errorf("buyer balance is insufficient for fee")
	}
	state.Outputs[0].Satoshis -= fee
	state.Inputs[0].UnlockingScript = script.NewFromBytes(nil)
	return state, nil
}

func BuildArbitratedPoolFinalState(input StateInput) (*tx.Transaction, error) {
	return BuildArbitratedPoolState(input)
}
